### PR TITLE
Fix #1463. Allow to expose multiple ports when running in debug mode

### LINKED ISSUE #1463: Expose multiple ports through --debug-port when start lambda in debug mode
### Describe your idea/feature/enhancement
I would like to have multiple debug ports to be exposed on start lambda in debug mode.
JetBrains Rider IDE debugger requires two ports to be exposed to work properly (the process set up two sockets). We would like to have an option to start dotnet core docker container with multiple ports available.

### Proposal
It would be great to be able to specify a multiple coma-separated ports in `--debug-port` option, e.g. `--debug-port 5600, 5601`.

Add details of how to add this to the product.

Things to consider:
1. Will this require any updates to the [SAM Spec](https://github.com/awslabs/serverless-application-model)
No.

### Additional Details
None.

### PR DESCRIPTION
**Pull request by sdubov**
*Issue #, if available:*
#1463

*Description of changes:*
* Add ability to specify multiple ports to be exposed from docker instance when running lambda in debug mode
* Update and add new tests to verify multiple debug ports

*Checklist:*

- [x] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [x] Write unit tests
- [x] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: designs/debug_mode_multiple_exposed_ports.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+What will be changed?
+---------------------
+SAM CLI has a ``--debug-port`` parameter that provide a port. This parameter is stored in DebugContext object.
+``DebugContext`` should store an array of ports instead of one single port. This array should be transformed
+into a map containing each stored port when passing to docker container arguments.
+
+Success criteria for the change
+-------------------------------
+All ports (comma separated) specified in ``--debug-port`` SAM CLI parameter should be exposed by docker container.
```
Comment (jfuss) [inline designs/debug_mode_multiple_exposed_ports.md]: Why comma separated? This seems like a good use-case for multiple options: `--debug-port 5600 --debug-port 5601`.

Comment (sdubov) [reply designs/debug_mode_multiple_exposed_ports.md]: Good point. Updated with `multiple=true` option parameter.

File: samcli/local/docker/lambda_container.py [hunk 2]
```diff
@@ -96,16 +96,21 @@ def _get_exposed_ports(debug_options):
         bind to same port both inside and outside the container ie. Runtime process is started in debug mode with
         at given port inside the container and exposed to the host machine at the same port
 
-        :param int debug_port: Optional, integer value of debug port
+        :param list(int) debug_port: Optional, List of integer values of debug ports
```
Comment (jfuss) [inline samcli/local/docker/lambda_container.py]: This doc is wrong (was wrong previously). The parameter passed in is `debug_options` not the `debug_port`. Could you update this to be correct?

File: samcli/local/docker/lambda_container.py [hunk 3]
```diff
@@ -96,16 +96,21 @@ def _get_exposed_ports(debug_options):
...
-        return {
-            # container port : host port
-            debug_options.debug_port: debug_options.debug_port
-        }
+        if not debug_options.debug_port:
+            return None
+
+        # container port : host port
+        ports_map = {}
+        for port in debug_options.debug_port:
```
Comment (jfuss) [inline samcli/local/docker/lambda_container.py]: Can you update the docs for `debug_options.debug_port` to reflect changes to the type?

File: samcli/commands/local/cli_common/options.py [hunk 4]
```diff
@@ -95,6 +95,8 @@ def invoke_common_options(f):
                 help="When specified, Lambda function container will start in debug mode and will expose this "
                 "port on localhost.",
                 envvar="SAM_DEBUG_PORT",
+                type=click.INT,
```
[the remark this answers is not in the document]
Comment (sdubov) [reply samcli/commands/local/cli_common/options.py]: No issues with passing integer as a string here. Looked to the `click` parameters type processing logic - it tries to case a value to integer. Verified it works well with strings unless you specify an invalid value, e.g. "abc" that should be an expected behavior.

Looked through `click` implementation for handling `multiple` flag - it wraps values into `tuple` if flag is set and return them or return an empty tuple if no values. So, I would expect we always have `tuple` after processing `--debug-port` option. Verified with single and multiple values.

File: samcli/local/docker/lambda_container.py (hunk 2 above)
Comment (sdubov) [reply samcli/local/docker/lambda_container.py]: Sure. Fixed doc here and in `DebugContext` to specify that debug port is a collection of ports to be exposed.

File: samcli/commands/local/cli_common/options.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jfuss) [reply samcli/commands/local/cli_common/options.py]: Perfect. Thanks for digging into that.

File: samcli/commands/local/lib/debug_context.py [hunk 5]
```diff
@@ -5,6 +5,13 @@
 
 class DebugContext(object):
     def __init__(self, debug_port=None, debugger_path=None, debug_args=None):
+        """
+        Initialize the Debug Context with Lambda debugger options
+
+        :param tuple(int) debug_port: Collection of debugger ports to be exposed from a docker container
```
Comment (jfuss) [inline samcli/commands/local/lib/debug_context.py]: Nit: Should this be `debug_ports` instead of `debug_port`

**Comment by jfuss** [comment]
@sdubov I had one last comment. It's purely for readability. Everything else looks good to me.

File: samcli/commands/local/lib/debug_context.py (hunk 5 above)
Comment (sdubov) [reply samcli/commands/local/lib/debug_context.py]: Agreed. I thought about the same naming change. My original thought here was to persist the original CLI option `--debug-port` down to usages. Updated parameter names from `InvokeContext` down to `DebugContext`.

Comment (jfuss) [reply samcli/commands/local/lib/debug_context.py]: Ya we can keep the option the same (`--debug-port`). This is just an internal object, so no concerns on the updates. 

Will do a re-review here shortly.

**Comment by sdubov** [comment]
@jfuss, there is a merge conflict in debug_context.py file. Would you like me to squash my commits into one and rebase? or merge `develop` into my branch is ok?

**Comment by sdubov** [comment]
Not sure sure `make-pr` configuration is failed. Checked agent details and did not find any issues from logs.

**Comment by jfuss** [comment]
@sdubov its failing due to a format check. We run a tool called black to make sure everything is formatted the same way throughout the codebase and avoids us having to do things manually. I can pull a update that part tomorrow if you don’t want to set that up.

**Comment by sdubov** [comment]
@jfuss, Thank you for clarifying. No worries. Done with code formatting.

File: samcli/local/docker/lambda_container.py [hunk 6]
```diff
@@ -169,17 +174,20 @@ def _get_entry_point(runtime, debug_options=None):  # pylint: disable=too-many-b
...
         :return list: List containing the new entry points. Each element in the list is one portion of the command.
             ie. if command is ``node index.js arg1 arg2``, then this list will be ["node", "index.js", "arg1", "arg2"]
         """
 
         if not debug_options:
             return None
 
-        debug_port = debug_options.debug_port
+        debug_ports = debug_options.debug_ports
+        if not debug_ports:
```
Comment (jfuss) [inline samcli/local/docker/lambda_container.py]: This check is fine but I think it is not needed. `__bool__` on DebugContext is `bool(self.debug_ports)`. This should never execute given the first if statement (`if not debug_options`).

**Comment by jfuss** [comment]
@sdubov Thanks for contributing the feature and going through and addressing all the feedback!

**Comment by sdubov** [comment]
@jfuss, no problem :) thank you for your feedback and a quick review 👍

**Pull request merged**